docs/user/implement_algo.rst

```diff
@@ -348,9 +358,107 @@ For completeness, the full experiment file is repeated below:
+
+.. code-block:: python
+
+    import tensorflow as tf
+    import numpy as np
+
+    from garage import log_performance, TrajectoryBatch
+    from garage.sampler import RaySampler
+    from garage.misc import tensor_utils
+
+    class MyAlgorithm():
+
+        sampler_cls = RaySampler
+
+        def __init__(self, env_spec, policy):
+            self.env_spec = env_spec
+            self.policy = policy
+            self.max_path_length = 200
+            self._lr = 1e-3
+            self._discount = 0.99
+
+            obs_space = self.policy.observation_space
+            action_space = self.policy.action_space
+            with tf.name_scope('inputs'):
+                self._obs = obs_space.to_tf_placeholder(name='obs', batch_dims=2)
+                self._action = action_space.to_tf_placeholder(name='action',
+                                                              batch_dims=2)
+                self._returns = tf.compat.v1.placeholder(tf.float32,
+                                                         shape=[None, None],
+                                                         name='return')
+            policy_dist = self.policy.build(self._obs, name='policy').dist
+            with tf.name_scope('loss'):
+                ll = policy_dist.log_prob(self._action, name='log_likelihood')
+                loss = -tf.reduce_mean(ll * self._returns)
+            with tf.name_scope('train'):
+                self._train_op = tf.compat.v1.train.AdamOptimizer(
+                    self._lr).minimize(loss)
+
+        def train(self, runner):
+            for epoch in runner.step_epochs():
+                samples = runner.obtain_samples(epoch)
+                log_performance(epoch,
+                                TrajectoryBatch.from_trajectory_list(self.env_spec,
+                                                                     samples),
+                                self._discount)
+                self.train_once(samples)
+
+        def train_once(self, paths):
+            obs = [path['observations'] for path in paths]
+            obs = tensor_utils.pad_tensor_n(obs, self.max_path_length)
+            actions = [path['actions'] for path in paths]
+            actions = tensor_utils.pad_tensor_n(actions, self.max_path_length)
+            returns = []
+            for path in paths:
+                returns.append(tensor_utils.discount_cumsum(path['rewards'],
+                                                           self._discount))
+            returns = tensor_utils.pad_tensor_n(returns, self.max_path_length)
+            sess = tf.compat.v1.get_default_session()
+            sess.run(self._train_op, feed_dict={
+                 self._obs: np.stack(obs, axis=0),
+                 self._action: np.stack(actions, axis=0),
+                 self._returns: np.stack(returns, axis=0),
+            })
+            return np.mean(returns)
+
+        # For pickling
+        def __getstate__(self):
```

Comment:
You should probably label this "For pickling / snapshot" and point to the other docs which are relevant (probably in the paragraph above this source block). We should really also have `__setstate__` here (probably using a method named `_init_opt` or something), to make the whole picture clear.